packages/server-wallet/src/wallet/index.ts

```diff
@@ -126,53 +126,22 @@ export class Wallet implements WalletInterface {
   }
 
   async pushMessage(message: AddressedMessage): Result {
-    const channelIds: Bytes32[] = [];
-
-    try {
-      await Channel.transaction(async tx => {
-        for (const ss of message.signedStates || []) {
-          // We ignore unsigned states
-          if (!ss.signatures?.length) return;
-
-          const channelId = calculateChannelId(ss);
-          let channel = await Channel.query(tx)
-            .where('channelId', channelId)
-            .first();
-
-          if (!channel) {
-            const addresses = ss.participants.map(p => p.signingAddress);
-            const signingWallet = await SigningWallet.query(tx)
-              .whereIn('address', addresses)
-              .first();
-
-            if (!signingWallet) {
-              logger.error(
-                {
-                  knownWallets: await SigningWallet.query(tx).select(),
-                  addresses,
-                },
-                'Not in channel'
-              );
-              throw Error('Not in channel');
-            }
-
-            const {address: signingAddress} = signingWallet;
-            const cols: RequiredColumns = {...ss, vars: [addHash(ss)], signingAddress};
-
-            channel = Channel.fromJson(cols);
-            const {channelId} = await Channel.query(tx).insert(channel);
-            channelIds.push(channelId);
-          } else {
-            ss.signatures?.map(sig => channel.addState(ss, sig));
-            await Channel.query(tx).update(channel);
-            channelIds.push(channel.channelId);
-          }
-        }
-      });
-    } catch (err) {
-      logger.error({err}, 'Could not push message');
-      throw err;
-    }
+    await Store.transaction(async () => {
+      if (message.signedStates) {
+        message.signedStates.forEach(async ss => {
+          await Store.addSignedState(ss);
```

Comment:
If one signature is invalid, should the entire API call return an error? Should it return one successful `ChannelUpdated` outbox notification and one `Error` notification? Also, since it is within a transaction, if one fails would they all fail to persist to disk?

Reply:
Currently the assumption is that if a signature is invalid or a state is malformed the whole `PushMessage` call fails and nothing is persisted. I created #2369 for further discussion around this.

I think a common case is receiving a message containing `SignedState`s and `Objective`s for a common task (ie: virtual funding), so it seems to make sense to either accept them all or reject them all instead of partially updating the store. It also makes it an easier API call for the `app` to understand, as it would either fail or succeed instead of partially succeeding/failing.

That said there may be some performance considerations that we might run into or other things to consider!

Reply:
It seems to me like if we have valid data that may at some point be useful, we should probably persist it just to be safe. Maybe this should be true for `signedStates` and not `objectives`?